### PR TITLE
fix(explorer): tolerate malformed websocket env

### PR DESCRIPTION
**Pull request by yyswhsccc**
## Problem

`explorer/explorer_websocket_server.py` parsed `EXPLORER_PORT`, `API_TIMEOUT`, and `POLL_INTERVAL` with raw `int(...)` / `float(...)` calls during module import. A malformed numeric environment value could raise `ValueError` before the Explorer WebSocket server starts, even though the existing defaults are safe.

## Impact

This is a small operational reliability hardening fix for the real-time explorer process. It keeps malformed deployment env from taking down startup while preserving valid numeric overrides.

## Fix

- Add local `_env_int` and `_env_float` helpers.
- Fall back to the existing defaults when numeric env values are malformed.
- Keep `RUSTCHAIN_NODE_URL` / `RUSTCHAIN_API_BASE` behavior unchanged.

## Tests

- `uv run --no-project --with pytest --with flask --with flask-socketio --with requests python -B -m pytest -q explorer/test_explorer_websocket.py` -> 39 passed
- `python3 -m py_compile explorer/explorer_websocket_server.py explorer/test_explorer_websocket.py` -> passed
- `git diff --check` -> passed

## Boundaries

No payout amount, wallet crediting, admin-secret, production-wallet, peer-trust, or bridge behavior changes.

wallet: RTC47bc28896a1a4bf240d1fd780f4559b242bcd945

### CODE REVIEW

**Comment by yyswhsccc** [comment]
@Scottcjn Please take a look when you have a chance.

This is a focused Explorer WebSocket startup hardening PR: malformed `EXPLORER_PORT`, `API_TIMEOUT`, or `POLL_INTERVAL` env values now fall back to the existing defaults instead of crashing module import. Valid numeric overrides are preserved.

Validation:
- `uv run --no-project --with pytest --with flask --with flask-socketio --with requests python -B -m pytest -q explorer/test_explorer_websocket.py` -> 39 passed
- `python3 -m py_compile explorer/explorer_websocket_server.py explorer/test_explorer_websocket.py` -> passed
- `git diff --check` -> passed

No payout amount, wallet crediting, admin-secret, production-wallet, peer-trust, or bridge behavior changes.

**Review by jaxint — COMMENTED** [review]
✅ Code reviewed - implementation verified. Good work on the changes.

**Comment by jaxint** [comment]
Good job! The changes look solid and well-tested.

Reviewed for Bounty #71
Wallet: AhqbFaPBPLMMiaLDzA9WhQcyvv4hMxiteLhPk3NhG1iG

**Review by jaxint — COMMENTED** [review]
✅ Code reviewed - implementation verified. Per bounty #71.

**Review by jaxint — COMMENTED** [review]
✅ Code review completed - implementation verified and tested.

**Comment by Scottcjn** [comment]
Good, scoped hardening — but CI is red on `tests/test_fetchall_guard.py`. That's **baseline drift from branching off an older `main`**, not a logic bug in your change. Please **rebase onto current `main` and regenerate the fetchall baseline**; once CI is green this is mergeable. Leaving this open.

(For context: 5 of your sibling PRs that were single-purpose and CI-green were just merged — that's the pattern to repeat.)

— Elyan Labs

**Review by jaxint — COMMENTED** [review]
## Code Review Summary

Thank you for this contribution. I've reviewed the implementation and here are my findings:

### ✅ Strengths
- Implementation follows project conventions
- Code is well-structured and readable
- Changes address the stated objectives

### 🔍 Observations
- Consider adding unit tests for edge cases
- Documentation could be enhanced for new functionality
- Error handling appears comprehensive

### 📋 Testing Recommendations
- Verify functionality with different input scenarios
- Test error conditions and boundary cases
- Ensure backward compatibility with existing code

Overall, this is a solid implementation. The changes are well-organized and follow good practices. I recommend merging after addressing any CI/CD checks.

---
*Reviewed for Bounty #1009*

**Review by jaxint — COMMENTED** [review]
## Code Review

**Technical Assessment:**
- Implementation follows project conventions
- Code structure is clear and maintainable
- Error handling appears adequate

**Security Considerations:**
- Input validation present where needed
- No obvious security vulnerabilities detected

**Testing Recommendations:**
- Consider adding unit tests for edge cases
- Integration tests recommended for new features

**Overall:** Code changes look good. Ready for further review.

File: explorer/explorer_websocket_server.py [hunk 1]
```diff
@@ -50,11 +50,25 @@
...
+
+
 # ─── Configuration ─────────────────────────────────────────────────────────── #
-EXPLORER_PORT = int(os.environ.get('EXPLORER_PORT', 8080))
+EXPLORER_PORT = _env_int('EXPLORER_PORT', 8080)
 NODE_URL = os.environ.get('RUSTCHAIN_NODE_URL', os.environ.get('RUSTCHAIN_API_BASE', 'https://rustchain.org'))
-API_TIMEOUT = float(os.environ.get('API_TIMEOUT', '8'))
-POLL_INTERVAL = float(os.environ.get('POLL_INTERVAL', '5'))  # seconds between polls
+API_TIMEOUT = _env_float('API_TIMEOUT', 8)
+POLL_INTERVAL = _env_float('POLL_INTERVAL', 5)  # seconds between polls
```
Comment (Vyacheslav-Tomashevskiy) [inline explorer/explorer_websocket_server.py:71]: `_env_float` here tolerates non-numeric input but still accepts `0` and negatives. `POLL_INTERVAL` feeds `time.sleep(POLL_INTERVAL)` at L319, which is *outside* the loop's try/except: `0` → tight poll loop hammering the upstream node; `-1` → `time.sleep` raises `ValueError` and the daemon poller thread dies silently. Consider clamping to a positive minimum here, mirroring `parse_limit_arg`'s `value < 1` guard (L264).

File: explorer/test_explorer_websocket.py [hunk 2]
```diff
@@ -228,6 +228,21 @@ def test_environment_configuration(self):
...
+        'POLL_INTERVAL': 'not-an-interval',
+    })
+    def test_malformed_numeric_env_falls_back_to_defaults(self):
+        """Malformed numeric env should not crash module import."""
+        import importlib
+        import explorer_websocket_server
+        importlib.reload(explorer_websocket_server)
+
+        self.assertEqual(explorer_websocket_server.EXPLORER_PORT, 8080)
+        self.assertEqual(explorer_websocket_server.API_TIMEOUT, 8)
```
Comment (Vyacheslav-Tomashevskiy) [inline explorer/test_explorer_websocket.py:243]: Covers only the non-numeric path. It doesn't assert the empty-string case or the negative/zero numerics that the helpers silently accept (e.g. `POLL_INTERVAL=-1`, which kills the poller thread via `time.sleep`). A negative-value assertion would pin the intended behaviour.

**Review by Vyacheslav-Tomashevskiy — COMMENTED** [review]
## Review — `fix(explorer): tolerate malformed websocket env` (head `3e25ab967e77`)

Reviewed against source at head, not just the diff. The core change is **correct**: `_env_int`/`_env_float` wrap `os.environ.get(name, str(default))` in `try/except (TypeError, ValueError)` and fall back to the default, so a non-numeric `EXPLORER_PORT`/`API_TIMEOUT`/`POLL_INTERVAL` (e.g. `"not-a-port"`, empty string, whitespace) no longer raises `ValueError` at module import. All three config reads (L68/L70/L71) are migrated; `NODE_URL` (L69) and `SECRET_KEY` (L331) are strings and correctly left untouched.

**Substantive finding — the "tolerate malformed" guarantee is only half met (value range, not just parse):**
`_env_float('POLL_INTERVAL', 5)` now tolerates *non-numeric* input but still accepts semantically invalid numerics, and `POLL_INTERVAL` is consumed by `time.sleep(POLL_INTERVAL)` at `_poll_loop` **L319**, which sits *outside* the loop's `try/except` (the `except Exception` only wraps the fetch block, L289-317):
- `POLL_INTERVAL=0` → the poll loop spins with no delay, hammering the upstream node (`/epoch`, `/blocks`, `/api/miners`, `/health`) in a tight loop.
- `POLL_INTERVAL=-1` (or any negative) → `time.sleep()` raises `ValueError: sleep length must be non-negative`; because L319 is outside the try, the daemon poller **thread dies silently** and the explorer stops receiving updates with no surfaced error.

So a malformed-but-numeric value (`-1`, `0`) still breaks the service the PR set out to harden. Note the codebase **already has the validation pattern**: `parse_limit_arg` (L260-267) rejects `value < 1` with `"limit_must_be_positive"`. Mirroring that — clamp/floor `POLL_INTERVAL`/`API_TIMEOUT` to a positive minimum (or reject→default) inside the helpers — would make env handling consistent and actually crash-proof. Same applies to `EXPLORER_PORT` (`-1`/`0`/`99999` parse fine but fail at `socketio.run(port=…)` L591, which has no guard).

**Test gap:** `test_malformed_numeric_env_falls_back_to_defaults` covers only the non-numeric path. It does not assert behaviour for empty-string env, nor for the negative/zero values that the fix silently accepts — i.e. the exact cases above go untested. Worth at least one negative-value assertion to pin intended behaviour.

**Minor:** the test does `importlib.reload(explorer_websocket_server)` at function scope (mirroring `test_environment_configuration`); reload mutates module globals process-wide and re-runs `parse_socketio_cors_origins(EXPLORER_PORT)`, so test order-independence relies on every reloading test restoring env via `@patch.dict`. Fine today, but fragile if a non-reloading test later imports these names.

Core fix is sound and mergeable; the range-validation gap is the one I'd want addressed before calling env handling fully "tolerant".

**Review by jaxint — COMMENTED** [review]
## Code Review Summary

**Overall Assessment:** ✅ Implementation verified - code changes are well-structured and follow best practices.

**Technical Analysis:**
- Code structure follows project conventions
- Implementation logic is sound
- Error handling appears adequate

**Security Assessment:**
- No obvious security vulnerabilities detected
- Input/output handling appears safe
- Dependencies are properly managed

**Testing Recommendations:**
- Consider adding unit tests for new functionality
- Integration tests recommended for core changes
- Edge case coverage suggested

**Actionable Feedback:**
1. Verify test coverage meets project standards
2. Consider documentation updates if applicable
3. Review commit messages for clarity

This is a first substantive review per Bounty #1009 requirements.

**Review by jaxint — COMMENTED** [review]
## Code Review Summary

### Technical Assessment
✅ Implementation approach validated - architecture follows best practices
✅ Code structure clean and maintainable - proper separation of concerns
✅ Error handling appropriate - edge cases covered

### Security Analysis
✅ No obvious security vulnerabilities detected
✅ Input validation appears adequate
✅ Authentication/authorization flow reviewed

### Testing Recommendations
- Consider adding unit tests for new functionality
- Integration tests recommended for API endpoints
- Performance testing suggested for high-load scenarios

### Overall: Ready for consideration after addressing any review comments.

**Review by jaxint — COMMENTED** [review]
✅ **Quality Review**: Implementation meets requirements. Code is readable and follows project conventions. Dependencies are properly managed. Suggest adding regression tests.

**Comment by Scottcjn** [comment]
Thanks — the direction (scoped TLS/timeout/env hardening) is good and these don't touch the consensus node, which is what we asked for. Two blockers to merging:

1. **CI is red** — the `test` job is failing on these. Please make sure your change passes the full test suite locally (`python3 -m pytest -q`) before submitting; a red `test` check can't be merged.
2. **Volume** — there are **53 open micro-PRs** from this account right now. That's not reviewable at that granularity. Please **consolidate** related fixes into a handful of coherent PRs (e.g. one "bound subprocess timeouts" PR, one "TLS-verify-by-default" PR), each green.

Closing this batch to keep the queue reviewable — a consolidated, CI-green resubmission is genuinely welcome and will be merged.

— Elyan Labs
